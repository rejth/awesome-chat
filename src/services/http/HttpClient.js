export default class HttpClient {
  fetch = fetch;

  constructor(fetchFunction) {
    this.fetch = fetchFunction || fetch;
  }

  get(url, data) {
    const urlWithData = HttpClient.getUrlWithData(url, data);
    return this.sendRequest('get', urlWithData);
  }

  post(url, body) {
    return this.sendRequest('post', url, body);
  }

  put(url, body) {
    return this.sendRequest('put', url, body);
  }

  delete(url, data) {
    const urlWithData = HttpClient.getUrlWithData(url, data);
    return this.sendRequest('delete', urlWithData, null);
  }

  sendRequest(method, url, data) {
    const options = HttpClient.getRequestOptions(method, url);
    const optionsWithBody = HttpClient.getOptionsWithBody(options, data);
    return this.fetch(optionsWithBody);
  }

  static getRequestOptions(method, url) {
    const accessToken = HttpClient.getAccessToken();
    if (accessToken) {
      return {
        url,
        method,
        headers: {
          'Content-Type': 'application/json;charset=utf-8',
          Authorization: `Bearer ${accessToken}`,
        },
      };
    }

    return {
      url,
      method,
      headers: {
        'Content-Type': 'application/json;charset=utf-8',
      },
    };
  }

  static getAccessToken() {
    const credentials = JSON.parse(localStorage.getItem('credentials'));
    if (!credentials || !credentials?.token) return null;
    return credentials.token;
  }

  static getOptionsWithBody(options, data) {
    return data
      ? { ...options, data }
      : options;
  }

  // при использовании Fetch API необходимо преобразовывать body в строку
  // и отправлять {...options, body: '{ key: value }'}
  static getDataLikeString(data) {
    if (!data) return data;
    if (typeof data === 'object') return JSON.stringify(data);
    if (typeof data.toString === 'function') return data.toString();
    return data;
  }

  static getUrlWithData(url, data) {
    if (!data || typeof data !== 'object') return url;

    const formattedUrl = url.includes('?') ? `${url}&` : `${url}?`;
    const parameters = Object
      .entries(data)
      .filter((item) => (item[1] || item[1] === 0 || item[1] === false))
      .map(([key, value]) => {
        const formattedValue = encodeURIComponent(`${value}`);
        return `${key}=${formattedValue}`;
      })
      .join('&');

    if (!parameters) return url;

    return `${formattedUrl}${parameters}`;
  }
}
