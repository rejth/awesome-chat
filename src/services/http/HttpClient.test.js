import {
  describe,
  beforeAll,
  test,
  jest,
  expect,
} from '@jest/globals';

import HttpClient from './HttpClient';

const http = new HttpClient((optionsWithBody) => (
  Promise.resolve(optionsWithBody)
));

HttpClient.getAccessToken = jest.fn();

describe('Testing an instance of the HttpClient class', () => {
  let method = '';
  let url = '';
  let data = {};

  beforeAll(() => {
    method = 'post';
    url = 'https://some-url.com';
    data = { first: 1, second: 2 };
  });

  test('getRequestOptions with authorization token', () => {
    HttpClient.getAccessToken.mockReturnValue('token');
    const options = HttpClient.getRequestOptions(method, url);

    expect(HttpClient.getAccessToken).toHaveBeenCalledTimes(1);
    expect(options?.method).toBe(method);
    expect(options?.url).toBe(url);
    expect(options).toEqual(expect.objectContaining({
      url: expect.any(String),
      method: expect.any(String),
      headers: expect.objectContaining({
        'Content-Type': expect.any(String),
        Authorization: 'Bearer token',
      }),
    }));
    expect(options).toMatchSnapshot();
  });

  test('getRequestOptions without authorization token', () => {
    HttpClient.getAccessToken.mockReturnValue(null);
    const options = HttpClient.getRequestOptions(method, url);

    expect(HttpClient.getAccessToken).toHaveBeenCalledTimes(1);
    expect(options).toEqual(expect.objectContaining({
      headers: expect.not.objectContaining({
        Authorization: expect.any(String),
      }),
    }));
    expect(options).toMatchSnapshot();
  });

  test('getOptionsWithBody', () => {
    HttpClient.getAccessToken.mockReturnValue('token');
    const spy = jest.spyOn(HttpClient, 'getRequestOptions');

    const options = HttpClient.getRequestOptions(method, url);
    const optionsWithBody = HttpClient.getOptionsWithBody(options, data);

    expect(HttpClient.getAccessToken).toHaveBeenCalledTimes(1);
    expect(spy).toHaveBeenCalledTimes(1);
    expect(optionsWithBody).toHaveProperty('data');
    expect(optionsWithBody?.data).toMatchObject(data);
    spy.mockRestore();
    expect(optionsWithBody).toMatchSnapshot();
  });

  test('getUrlWithData', () => {
    const result = HttpClient.getUrlWithData(url, data);
    expect(result).toBe(`${url}?first=1&second=2`);
    expect(result).toMatchSnapshot();
  });

  test('getDataLikeString', () => {
    expect(HttpClient.getDataLikeString(data)).toBe(JSON.stringify(data));
    expect(HttpClient.getDataLikeString(null)).toBeNull();
  });

  test('method GET', async () => {
    HttpClient.getAccessToken.mockReturnValue('some-token');
    const result = await http.get(url, data);

    expect(HttpClient.getAccessToken).toHaveBeenCalledTimes(1);
    expect(result?.url).toBe(`${url}?first=1&second=2`);
    expect(result).not.toHaveProperty('data');
    expect(result).toMatchSnapshot();
  });

  test('method POST', async () => {
    HttpClient.getAccessToken.mockReturnValue('some-token');
    const result = await http.post(url, data);

    expect(HttpClient.getAccessToken).toHaveBeenCalledTimes(1);
    expect(result?.url).toBe(url);
    expect(result).toHaveProperty('data');
    expect(result?.data).toMatchObject(data);
    expect(result).toMatchSnapshot();
  });

  test('method PUT', async () => {
    HttpClient.getAccessToken.mockReturnValue('some-token');
    const result = await http.put(url, data);

    expect(HttpClient.getAccessToken).toHaveBeenCalledTimes(1);
    expect(result?.url).toBe(url);
    expect(result).toHaveProperty('data');
    expect(result?.data).toMatchObject(data);
    expect(result).toMatchSnapshot();
  });

  test('method DELETE', async () => {
    HttpClient.getAccessToken.mockReturnValue('some-token');
    const result = await http.delete(url, data);

    expect(HttpClient.getAccessToken).toHaveBeenCalledTimes(1);
    expect(result?.url).toBe(`${url}?first=1&second=2`);
    expect(result).not.toHaveProperty('data');
    expect(result).toMatchSnapshot();
  });
});
