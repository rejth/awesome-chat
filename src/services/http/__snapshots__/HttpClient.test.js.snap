// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing an instance of the HttpClient class getOptionsWithBody 1`] = `
Object {
  "data": Object {
    "first": 1,
    "second": 2,
  },
  "headers": Object {
    "Authorization": "Bearer token",
    "Content-Type": "application/json;charset=utf-8",
  },
  "method": "post",
  "url": "https://some-url.com",
}
`;

exports[`Testing an instance of the HttpClient class getRequestOptions with authorization token 1`] = `
Object {
  "headers": Object {
    "Authorization": "Bearer token",
    "Content-Type": "application/json;charset=utf-8",
  },
  "method": "post",
  "url": "https://some-url.com",
}
`;

exports[`Testing an instance of the HttpClient class getRequestOptions without authorization token 1`] = `
Object {
  "headers": Object {
    "Content-Type": "application/json;charset=utf-8",
  },
  "method": "post",
  "url": "https://some-url.com",
}
`;

exports[`Testing an instance of the HttpClient class getUrlWithData 1`] = `"https://some-url.com?first=1&second=2"`;

exports[`Testing an instance of the HttpClient class method DELETE 1`] = `
Object {
  "headers": Object {
    "Authorization": "Bearer some-token",
    "Content-Type": "application/json;charset=utf-8",
  },
  "method": "delete",
  "url": "https://some-url.com?first=1&second=2",
}
`;

exports[`Testing an instance of the HttpClient class method GET 1`] = `
Object {
  "headers": Object {
    "Authorization": "Bearer some-token",
    "Content-Type": "application/json;charset=utf-8",
  },
  "method": "get",
  "url": "https://some-url.com?first=1&second=2",
}
`;

exports[`Testing an instance of the HttpClient class method POST 1`] = `
Object {
  "data": Object {
    "first": 1,
    "second": 2,
  },
  "headers": Object {
    "Authorization": "Bearer some-token",
    "Content-Type": "application/json;charset=utf-8",
  },
  "method": "post",
  "url": "https://some-url.com",
}
`;

exports[`Testing an instance of the HttpClient class method PUT 1`] = `
Object {
  "data": Object {
    "first": 1,
    "second": 2,
  },
  "headers": Object {
    "Authorization": "Bearer some-token",
    "Content-Type": "application/json;charset=utf-8",
  },
  "method": "put",
  "url": "https://some-url.com",
}
`;
